experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyField.ts

```diff
@@ -0,0 +1,438 @@
+	FieldSchema,
+	TreeSchema,
+	schemaIsFieldNode,
+	schemaIsLeaf,
+} from "../typed-schema";
+import { TreeStatus, treeStatusFromPath } from "../editable-tree";
+import { Context } from "./context";
+import {
+	FlexibleNodeContent,
+	OptionalField,
+	Sequence,
+	TypedField,
+	TypedNodeUnion,
+	UnboxField,
+	UnboxNode,
+	UnboxNodeUnion,
+	UntypedField,
+	UntypedTree,
+	ValueField,
+} from "./editableTreeTypes";
+import { makeTree } from "./lazyTree";
+import {
+	LazyEntity,
+	anchorSymbol,
+	cursorSymbol,
+	forgetAnchorSymbol,
+	isFreedSymbol,
+	makePropertyEnumerableOwn,
+	makePropertyNotEnumerable,
+	tryMoveCursorToAnchorSymbol,
+} from "./lazyEntity";
+
+export function makeField(
+	context: Context,
+	schema: FieldSchema,
+	cursor: ITreeSubscriptionCursor,
+): UntypedField {
+	const fieldAnchor = cursor.buildFieldAnchor();
+
+	const field = new (kindToClass.get(schema.kind) ?? fail("missing field implementation"))(
+		context,
+		schema,
+		cursor,
+		fieldAnchor,
+	);
+
+	// Fields currently live as long as their parent does.
+	// For root fields, this means forever, but other cases can be cleaned up when their parent anchor is deleted.
+	if (fieldAnchor.parent !== undefined) {
+		const anchorNode =
+			context.forest.anchors.locate(fieldAnchor.parent) ??
+			fail("parent anchor node should always exist since field is under a node");
+		anchorNode.on("afterDelete", () => {
+			field[disposeSymbol]();
+		});
+	}
+	return field;
+}
+
+/**
+ * A Proxy target, which together with a `fieldProxyHandler` implements a basic access to
+ * the nodes of {@link EditableField} by means of the cursors.
+ */
+export abstract class LazyField<TKind extends FieldKindTypes, TTypes extends AllowedTypes>
+	extends LazyEntity<FieldSchema<TKind, TTypes>, FieldAnchor>
+	implements UntypedField
+{
+	public readonly key: FieldKey;
+
+	public constructor(
+		context: Context,
+		schema: FieldSchema<TKind, TTypes>,
+		cursor: ITreeSubscriptionCursor,
+		fieldAnchor: FieldAnchor,
+	) {
+		super(context, schema, cursor, fieldAnchor);
+		assert(cursor.mode === CursorLocationType.Fields, "must be in fields mode");
+		this.key = cursor.getFieldKey();
+
+		makePropertyNotEnumerable(this, "key");
+	}
+
+	public is<TSchema extends FieldSchema>(schema: TSchema): this is TypedField<TSchema> {
+		assert(
+			this.context.schema.policy.fieldKinds.get(schema.kind.identifier) === schema.kind,
+			"Narrowing must be done to a kind that exists in this context",
+		);
+
+		if (schema.kind !== schema.kind) {
+			return false;
+		}
+		if (schema.types === undefined) {
+			return this.schema.types === undefined;
+		}
+		if (this.schema.types === undefined) {
+			return false;
+		}
+		return compareSets({ a: this.schema.types, b: schema.types });
+	}
+
+	public isSameAs(other: UntypedField): boolean {
+		assert(
+			other.context === this.context,
+			"Content from different editable trees should not be used together",
+		);
+		return this.key === other.key && this.parent === other.parent;
```

Comment:
there is a 1:1 relationship between editable tree nodes and logical document nodes. You cannot get two tree node objects for the same logical node: the caching on anchor nodes ensures this.

It would be nice to ensure this for fields as well, but is hard to do because of potentially infinate numbers of empty fields under maps, so caching them could leak memory. With this new API and schema2 it will be easier to fix that (but we would still want file cache using weak refs, and have to avoid even that leaking due to the withAnchor list on the context retaining them all anyway.... is hard to fix.

Reply:
That makes sense - good explanation, thanks. If nodes are never duplicated, then checking the parent and key of a field is sufficient for equality.

Reply:
Should we add documentation to this effect in the code?